File: swift/Sources/AetherNetBitTorrent/Metainfo.swift
// SPDX-License-Identifier: MIT

import Foundation

/// One file within a torrent: its path components and length.
public struct TorrentFileEntry {
    public let path: [String]
    public let length: Int64

    public init(path: [String], length: Int64) {
        self.path = path
        self.length = length
    }

    /// Path components joined with '/'.
    public var joinedPath: String { path.joined(separator: "/") }
}

/// A parsed BitTorrent v1 metainfo (.torrent). `infoHashV1` is the SHA-1 of the RAW
/// bencoded info dictionary as it appears in the file (not a re-encode), extracted by
/// byte-offset, so it matches real clients byte-for-byte.
public struct TorrentMetainfo {
    public let root: BDict
    public let info: BDict
    public let infoHashV1: [UInt8]      // 20 bytes
    public let name: String
    public let pieceLength: Int64
    public let pieceHashes: [[UInt8]]   // each 20-byte SHA-1
    public let files: [TorrentFileEntry]
    public let totalLength: Int64
    public let announceURLs: [String]
    public let isSingleFile: Bool

    /// Lowercase hex of `infoHashV1` (40 chars).
    public var infoHashV1Hex: String { BTHex.encode(infoHashV1) }
}

public enum MetainfoError: Error, Equatable {
    case malformed(String)
}

/// Parses .torrent bytes.
public func parseTorrent(_ data: [UInt8]) throws -> TorrentMetainfo {
    let rootVal = try bencodeDecode(data)
    let root = try rootVal.dictValue()

    guard let infoVal = root.get("info") else {
        throw MetainfoError.malformed("metainfo has no 'info' dictionary")
    }
    let info = try infoVal.dictValue()

    let infoSpan = try extractInfoSpan(data)
    let infoHash = BTHash.sha1(infoSpan)

    guard let nameVal = info.get("name") else {
        throw MetainfoError.malformed("info has no 'name'")
    }
    let name = try nameVal.textValue()

    guard let plVal = info.get("piece length") else {
        throw MetainfoError.malformed("info has no 'piece length'")
    }
    let pieceLength = try plVal.intValue()
    if pieceLength <= 0 {
        throw MetainfoError.malformed("'piece length' must be positive")
    }

    guard let piecesVal = info.get("pieces") else {
        throw MetainfoError.malformed("info has no 'pieces'")
    }
    let piecesBytes = try piecesVal.bytesValue()
    if piecesBytes.count % 20 != 0 {
        throw MetainfoError.malformed("'pieces' length \(piecesBytes.count) is not a multiple of 20")
    }
    var pieceHashes: [[UInt8]] = []
    pieceHashes.reserveCapacity(piecesBytes.count / 20)
    var pi = 0
    while pi < piecesBytes.count {
        pieceHashes.append(Array(piecesBytes[pi..<(pi + 20)]))
        pi += 20
    }

    var files: [TorrentFileEntry] = []
    var total: Int64 = 0
    var singleFile = false
    if let filesVal = info.get("files") {
        let list = try filesVal.listValue()
        for f in list {
            let fd = try f.dictValue()
            guard let lenVal = fd.get("length") else {
                throw MetainfoError.malformed("file entry has no 'length'")
            }
            let length = try lenVal.intValue()
            guard let pathVal = fd.get("path") else {
                throw MetainfoError.malformed("file entry has no 'path'")
            }
            let pathList = try pathVal.listValue()
            var parts: [String] = []
            parts.reserveCapacity(pathList.count)
            for p in pathList {
                parts.append(try p.textValue())
            }
            if parts.isEmpty {
                throw MetainfoError.malformed("file entry has an empty 'path'")
            }
            files.append(TorrentFileEntry(path: parts, length: length))
            total += length
        }
    } else {
        singleFile = true
        guard let lenVal = info.get("length") else {
            throw MetainfoError.malformed("single-file info has neither 'length' nor 'files'")
        }
        let length = try lenVal.intValue()
        files.append(TorrentFileEntry(path: [name], length: length))
        total = length
    }

    // Trackers: announce + announce-list, de-duplicated, order preserved.
    var announce: [String] = []
    var seen = Set<String>()
    func addTracker(_ u: String) {
        if !u.isEmpty && !seen.contains(u) {
            seen.insert(u)
            announce.append(u)
        }
    }
    if let a = root.get("announce"), let s = try? a.textValue() {
        addTracker(s)
    }
    if let al = root.get("announce-list"), let tiers = try? al.listValue() {
        for tier in tiers {
            if let ts = try? tier.listValue() {
                for t in ts {
                    if let s = try? t.textValue() {
                        addTracker(s)
                    }
                }
            }
        }
    }

    return TorrentMetainfo(
        root: root,
        info: info,
        infoHashV1: infoHash,
        name: name,
        pieceLength: pieceLength,
        pieceHashes: pieceHashes,
        files: files,
        totalLength: total,
        announceURLs: announce,
        isSingleFile: singleFile
    )
}

/// Returns the raw bencoded bytes of the top-level "info" value by walking the
/// dictionary with byte-offset tracking (structure already validated by parseTorrent).
func extractInfoSpan(_ data: [UInt8]) throws -> [UInt8] {
    if data.isEmpty || data[0] != UInt8(ascii: "d") {
        throw MetainfoError.malformed("metainfo is not a bencoded dictionary")
    }
    var pos = 1
    let infoKey = Array("info".utf8)
    while pos < data.count && data[pos] != UInt8(ascii: "e") {
        let (keyVal, keyEnd) = try bencodeDecodeN(data, pos)
        guard case .bytes(let key) = keyVal else {
            throw MetainfoError.malformed("dictionary key is not a byte string")
        }
        pos = keyEnd
        let valStart = pos
        let (_, valEnd) = try bencodeDecodeN(data, pos)
        pos = valEnd
        if key == infoKey {
            return Array(data[valStart..<valEnd])
        }
    }
    throw MetainfoError.malformed("metainfo has no 'info' key")
}

/// Creates single-file .torrent bytes for `data`, splitting into `pieceLength`-byte
/// pieces and SHA-1-hashing each. Byte-identical to the C# TorrentBuilder and Go
/// BuildSingleFileTorrent.
public func buildSingleFileTorrent(
    name: String,
    data: [UInt8],
    pieceLength: Int,
    announce: String
) throws -> [UInt8] {
    if name.isEmpty {
        throw MetainfoError.malformed("name is required")
    }
    if pieceLength <= 0 {
        throw MetainfoError.malformed("piece length must be positive")
    }
    let pieceCount = (data.count + pieceLength - 1) / pieceLength
    var pieces = [UInt8](repeating: 0, count: pieceCount * 20)
    for i in 0..<pieceCount {
        let start = i * pieceLength
        let end = min(start + pieceLength, data.count)
        let h = BTHash.sha1(Array(data[start..<end]))
        for j in 0..<20 {
            pieces[i * 20 + j] = h[j]
        }
    }

    let info = BDict()
    try info.add("length", .int(Int64(data.count)))
    try info.add("name", .text(name))
    try info.add("piece length", .int(Int64(pieceLength)))
    try info.add("pieces", .bytes(pieces))

    let root = BDict()
    if !announce.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        try root.add("announce", .text(announce))
    }
    try root.add("info", .dict(info))
    return bencodeEncode(.dict(root))
}

File: go/bittorrent/metainfo.go
// SPDX-License-Identifier: MIT

package bittorrent

import (
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"strings"
)

// TorrentFileEntry is one file within a torrent: its path components and length.
type TorrentFileEntry struct {
	Path   []string
	Length int64
}

// JoinedPath returns the path components joined with '/'.
func (e TorrentFileEntry) JoinedPath() string { return strings.Join(e.Path, "/") }

// TorrentMetainfo is a parsed BitTorrent v1 metainfo (.torrent). InfoHashV1 is the
// SHA-1 of the RAW bencoded info dictionary as it appears in the file (not a re-encode),
// so it matches real clients byte-for-byte.
type TorrentMetainfo struct {
	Root         *BDict
	Info         *BDict
	InfoHashV1   [20]byte
	Name         string
	PieceLength  int64
	PieceHashes  [][]byte // each 20-byte SHA-1
	Files        []TorrentFileEntry
	TotalLength  int64
	AnnounceURLs []string
	IsSingleFile bool
}

// InfoHashV1Hex is the lowercase hex of InfoHashV1 (40 chars).
func (m *TorrentMetainfo) InfoHashV1Hex() string { return hex.EncodeToString(m.InfoHashV1[:]) }

// ParseTorrent parses .torrent bytes.
func ParseTorrent(data []byte) (*TorrentMetainfo, error) {
	rootVal, err := Decode(data)
	if err != nil {
		return nil, err
	}
	root, err := AsDict(rootVal)
	if err != nil {
		return nil, err
	}
	infoVal, ok := root.Get("info")
	if !ok {
		return nil, fmt.Errorf("metainfo has no 'info' dictionary")
	}
	info, err := AsDict(infoVal)
	if err != nil {
		return nil, err
	}

	infoSpan, err := extractInfoSpan(data)
	if err != nil {
		return nil, err
	}
	infoHash := sha1.Sum(infoSpan)

	nameVal, ok := info.Get("name")
	if !ok {
		return nil, fmt.Errorf("info has no 'name'")
	}
	name, err := AsText(nameVal)
	if err != nil {
		return nil, err
	}

	plVal, ok := info.Get("piece length")
	if !ok {
		return nil, fmt.Errorf("info has no 'piece length'")
	}
	pieceLength, err := AsInt(plVal)
	if err != nil {
		return nil, err
	}
	if pieceLength <= 0 {
		return nil, fmt.Errorf("'piece length' must be positive")
	}

	piecesVal, ok := info.Get("pieces")
	if !ok {
		return nil, fmt.Errorf("info has no 'pieces'")
	}
	piecesBytes, err := AsBytes(piecesVal)
	if err != nil {
		return nil, err
	}
	if len(piecesBytes)%20 != 0 {
		return nil, fmt.Errorf("'pieces' length %d is not a multiple of 20", len(piecesBytes))
	}
	pieceHashes := make([][]byte, 0, len(piecesBytes)/20)
	for i := 0; i < len(piecesBytes); i += 20 {
		h := make([]byte, 20)
		copy(h, piecesBytes[i:i+20])
		pieceHashes = append(pieceHashes, h)
	}

	var files []TorrentFileEntry
	var total int64
	singleFile := false
	if filesVal, ok := info.Get("files"); ok {
		list, err := AsList(filesVal)
		if err != nil {
			return nil, err
		}
		for _, f := range list {
			fd, err := AsDict(f)
			if err != nil {
				return nil, err
			}
			lenVal, ok := fd.Get("length")
			if !ok {
				return nil, fmt.Errorf("file entry has no 'length'")
			}
			length, err := AsInt(lenVal)
			if err != nil {
				return nil, err
			}
			pathVal, ok := fd.Get("path")
			if !ok {
				return nil, fmt.Errorf("file entry has no 'path'")
			}
			pathList, err := AsList(pathVal)
			if err != nil {
				return nil, err
			}
			parts := make([]string, 0, len(pathList))
			for _, p := range pathList {
				s, err := AsText(p)
				if err != nil {
					return nil, err
				}
				parts = append(parts, s)
			}
			if len(parts) == 0 {
				return nil, fmt.Errorf("file entry has an empty 'path'")
			}
			files = append(files, TorrentFileEntry{Path: parts, Length: length})
			total += length
		}
	} else {
		singleFile = true
		lenVal, ok := info.Get("length")
		if !ok {
			return nil, fmt.Errorf("single-file info has neither 'length' nor 'files'")
		}
		length, err := AsInt(lenVal)
		if err != nil {
			return nil, err
		}
		files = append(files, TorrentFileEntry{Path: []string{name}, Length: length})
		total = length
	}

	// Trackers: announce + announce-list, de-duplicated, order preserved.
	var announce []string
	seen := map[string]bool{}
	add := func(u string) {
		if u != "" && !seen[u] {
			seen[u] = true
			announce = append(announce, u)
		}
	}
	if a, ok := root.Get("announce"); ok {
		if s, err := AsText(a); err == nil {
			add(s)
		}
	}
	if al, ok := root.Get("announce-list"); ok {
		if tiers, err := AsList(al); err == nil {
			for _, tier := range tiers {
				if ts, err := AsList(tier); err == nil {
					for _, t := range ts {
						if s, err := AsText(t); err == nil {
							add(s)
						}
					}
				}
			}
		}
	}

	m := &TorrentMetainfo{
		Root:         root,
		Info:         info,
		Name:         name,
		PieceLength:  pieceLength,
		PieceHashes:  pieceHashes,
		Files:        files,
		TotalLength:  total,
		AnnounceURLs: announce,
		IsSingleFile: singleFile,
	}
	copy(m.InfoHashV1[:], infoHash[:])
	return m, nil
}

// extractInfoSpan returns the raw bencoded bytes of the top-level "info" value by
// walking the dictionary with byte-offset tracking (structure already validated).
func extractInfoSpan(data []byte) ([]byte, error) {
	if len(data) == 0 || data[0] != 'd' {
		return nil, fmt.Errorf("metainfo is not a bencoded dictionary")
	}
	pos := 1
	for pos < len(data) && data[pos] != 'e' {
		keyVal, keyN, err := DecodeN(data[pos:])
		if err != nil {
			return nil, err
		}
		key, err := AsBytes(keyVal)
		if err != nil {
			return nil, fmt.Errorf("dictionary key is not a byte string")
		}
		pos += keyN
		valStart := pos
		_, valN, err := DecodeN(data[pos:])
		if err != nil {
			return nil, err
		}
		valEnd := pos + valN
		pos = valEnd
		if string(key) == "info" {
			return data[valStart:valEnd], nil
		}
	}
	return nil, fmt.Errorf("metainfo has no 'info' key")
}

// BuildSingleFileTorrent creates single-file .torrent bytes for data, splitting into
// pieceLength-byte pieces and SHA-1-hashing each. Byte-identical to the C# TorrentBuilder.
func BuildSingleFileTorrent(name string, data []byte, pieceLength int, announce string) ([]byte, error) {
	if name == "" {
		return nil, fmt.Errorf("name is required")
	}
	if pieceLength <= 0 {
		return nil, fmt.Errorf("piece length must be positive")
	}
	pieceCount := (len(data) + pieceLength - 1) / pieceLength
	pieces := make([]byte, pieceCount*20)
	for i := 0; i < pieceCount; i++ {
		start := i * pieceLength
		end := start + pieceLength
		if end > len(data) {
			end = len(data)
		}
		h := sha1.Sum(data[start:end])
		copy(pieces[i*20:], h[:])
	}

	info := NewBDict()
	_ = info.Add("length", BInt(int64(len(data))))
	_ = info.Add("name", BStr(name))
	_ = info.Add("piece length", BInt(int64(pieceLength)))
	_ = info.Add("pieces", BStr(pieces))

	root := NewBDict()
	if strings.TrimSpace(announce) != "" {
		_ = root.Add("announce", BStr(announce))
	}
	_ = root.Add("info", info)
	return Encode(root), nil
}
